src/electos/ballotmaker/cli.py

```diff
@@ -45,8 +51,8 @@ def make(
     """Make ballots from EDF file"""
     make_ballots_result = make_ballots.make_ballots(edf, output_dir, style)
     if make_ballots_result != NO_ERRORS:
-        typer.echo(
-            f"Error {make_ballots_result} making ballots: {strerror(make_ballots_result)}"
+        log.error(
+            f"Code {make_ballots_result} in make - {strerror(make_ballots_result)}"
```

Comment:
Do you really want to log here instead of `echo`?

In general I think yes: for a non-CLI application (or any library inside an application) it makes sense that errors always be sent to a logger, since they may not be run with a console. But a Typer `command` is explicitly in a CLI, where output is intended to be user visible on the console. It's the one use case the [Python logging tutorial](https://docs.python.org/3.8/howto/logging.html#basic-logging-tutorial) says to use `print` (or in this case `echo`) for. As as a secondary point by doing this you are explicitly giving up the formatting features of Typer.)

I'm not arguing you _shouldn't_ log here. There's advantage to uniformity. It just seems like one of the cases where it's not the right fit.

Reply:
I haven't decided yet, so this change was made in the spirit of "let's see how it works to use a unified notifcation system, compared to the benefits of Typer's `echo`. I think, as with the previous comment, my needs in development mode are different from what we'll want to see in the shipping version. 

Another benefit to consolidate all notifications onto `logger`: I can send logger output to the console when I'm running `pytest`, like this:

```
pytest --log-format="%(asctime)s %(levelname)s %(message)s" --log-cli-level=debug
```

This is incredibly useful because:

* pytest usually supresses console output, so now my tests give me valuable info on the console besides just pass/fail and coverage info.
* I have control over logging on a per-test basis (I can also use a `pytest.ini` file, as well as the command line options, to set options globally for all tests)
* This works on individual python modules -- in other words, I can call `logger` in a module without having to configure the logger object in the code, also without having to execute the main module to set up logging (so I don't have to run, `cli.py`)

So, as with the previous comment, I'm going to create an issue to track this, but leave the code as-is for now. My thoughts are that I'd like to go with Typer's `echo` -- or even `secho` -- for the production code.

Reply:
The definitive reference: [[Logging — pytest documentation](https://docs.pytest.org/en/6.2.x/logging.html)](https://docs.pytest.org/en/6.2.x/logging.html)

Reply:
Issue posted here: [[Use Typer's echo for output in cli.py · Issue #70 · TrustTheVote-Project/BallotLab](https://github.com/TrustTheVote-Project/BallotLab/issues/70)](https://github.com/TrustTheVote-Project/BallotLab/issues/70)